File: backend/src/main/java/com/zennyt/games/domain/model/GameRuntimeSnapshot.java
package com.zennyt.games.domain.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/** Immutable, non-scoring runtime configuration captured when a session starts. */
public record GameRuntimeSnapshot(
        UUID bankId,
        String bankCode,
        Integer bankVersion,
        String bankContentType,
        Integer settingsVersion,
        Integer modifiersVersion,
        Map<String, Object> settings,
        Map<String, Object> modifiers) {

    public GameRuntimeSnapshot {
        settings = immutableMap(settings);
        modifiers = immutableMap(modifiers);
    }

    public GameRuntimeSnapshot(Integer settingsVersion, Integer modifiersVersion,
                               Map<String, Object> settings,
                               Map<String, Object> modifiers) {
        this(null, null, null, null, settingsVersion, modifiersVersion, settings, modifiers);
    }

    public static GameRuntimeSnapshot empty() {
        return new GameRuntimeSnapshot(null, null, null, null,
            null, null, Map.of(), Map.of());
    }

    public int settingInt(String key, int fallback, int minimum, int maximum) {
        Object value = settings.get(key);
        if (!(value instanceof Number number)) return fallback;
        return Math.max(minimum, Math.min(maximum, number.intValue()));
    }

    public String settingString(String key, String fallback) {
        Object value = settings.get(key);
        return value instanceof String text && !text.isBlank() ? text : fallback;
    }

    private static Map<String, Object> immutableMap(Map<String, Object> source) {
        if (source == null || source.isEmpty()) return Map.of();
        Map<String, Object> copy = new LinkedHashMap<>();
        source.forEach((key, value) -> copy.put(key, immutableValue(value)));
        return Collections.unmodifiableMap(copy);
    }

    private static Object immutableValue(Object value) {
        if (value instanceof Map<?, ?> nested) {
            Map<String, Object> copy = new LinkedHashMap<>();
            nested.forEach((key, item) -> {
                if (!(key instanceof String text)) {
                    throw new IllegalArgumentException("Clé JSON runtime non textuelle");
                }
                copy.put(text, immutableValue(item));
            });
            return Collections.unmodifiableMap(copy);
        }
        if (value instanceof List<?> nested) {
            List<Object> copy = new ArrayList<>(nested.size());
            nested.forEach(item -> copy.add(immutableValue(item)));
            return Collections.unmodifiableList(copy);
        }
        return value;
    }
}

File: mobile/lib/features/games/domain/entities/game_runtime_snapshot.dart
/// Immutable non-scoring controls captured by the backend when a session starts.
class GameRuntimeSnapshot {
  const GameRuntimeSnapshot({
    this.bankId,
    this.bankCode,
    this.bankVersion,
    this.bankContentType,
    this.settingsVersion,
    this.modifiersVersion,
    this.settings = const {},
    this.modifiers = const {},
  });

  final String? bankId;
  final String? bankCode;
  final int? bankVersion;
  final String? bankContentType;
  final int? settingsVersion;
  final int? modifiersVersion;
  final Map<String, dynamic> settings;
  final Map<String, dynamic> modifiers;

  factory GameRuntimeSnapshot.fromJson(Map<String, dynamic>? json) {
    if (json == null) return const GameRuntimeSnapshot();
    return GameRuntimeSnapshot(
      bankId: json['bankId'] as String?,
      bankCode: json['bankCode'] as String?,
      bankVersion: (json['bankVersion'] as num?)?.toInt(),
      bankContentType: json['bankContentType'] as String?,
      settingsVersion: (json['settingsVersion'] as num?)?.toInt(),
      modifiersVersion: (json['modifiersVersion'] as num?)?.toInt(),
      settings: _freezeMap(json['settings'] as Map<String, dynamic>?),
      modifiers: _freezeMap(json['modifiers'] as Map<String, dynamic>?),
    );
  }

  bool settingBool(String key, {required bool fallback}) =>
      settings[key] is bool ? settings[key] as bool : fallback;

  bool modifierBool(String key, {required bool fallback}) =>
      modifiers[key] is bool ? modifiers[key] as bool : fallback;

  int modifierInt(
    String key, {
    required int fallback,
    required int minimum,
    required int maximum,
  }) {
    final value = modifiers[key];
    if (value is! num) return fallback;
    return value.toInt().clamp(minimum, maximum);
  }
}

Map<String, dynamic> _freezeMap(Map<String, dynamic>? source) =>
    Map<String, dynamic>.unmodifiable(
      (source ?? const {}).map(
        (key, value) => MapEntry(key, _freezeValue(value)),
      ),
    );

dynamic _freezeValue(dynamic value) {
  if (value is Map<String, dynamic>) return _freezeMap(value);
  if (value is List<dynamic>) {
    return List<dynamic>.unmodifiable(value.map(_freezeValue));
  }
  return value;
}
